netbsd/NetBSDProcessList.c

```diff
@@ -0,0 +1,311 @@
+   }
+
+   *s = '\0';
+
+   for (int i = 0; arg[i] != NULL; i++) {
+      size_t n = strlcat(s, arg[i], len);
+      if (i == 0) {
+         *basenameEnd = MINIMUM(n, len - 1);
+      }
+      /* the trailing space should get truncated anyway */
+      strlcat(s, " ", len);
+   }
+
+   return s;
+}
+
+/*
+ * Borrowed with modifications from NetBSD's top(1).
+ */
+static double getpcpu(const struct kinfo_proc2* kp) {
+   if (fscale == 0)
+      return 0.0;
+
+   return 100.0 * (double)kp->p_pctcpu / fscale;
+}
+
+static void NetBSDProcessList_scanProcs(NetBSDProcessList* this) {
+   const Settings* settings = this->super.settings;
+   bool hideKernelThreads = settings->hideKernelThreads;
+   bool hideUserlandThreads = settings->hideUserlandThreads;
+   int count = 0;
+   int nlwps = 0;
+
+   const struct kinfo_proc2* kprocs = kvm_getproc2(this->kd, KERN_PROC_ALL, 0, sizeof(struct kinfo_proc2), &count);
+
+   for (int i = 0; i < count; i++) {
+      const struct kinfo_proc2* kproc = &kprocs[i];
+
+      bool preExisting = false;
+      Process* proc = ProcessList_getProcess(&this->super, kproc->p_pid, &preExisting, NetBSDProcess_new);
+
+      proc->show = ! ((hideKernelThreads && Process_isKernelThread(proc)) || (hideUserlandThreads && Process_isUserlandThread(proc)));
+
+      if (!preExisting) {
+         proc->ppid = kproc->p_ppid;
+         proc->tpgid = kproc->p_tpgid;
+         proc->tgid = kproc->p_pid;
+         proc->session = kproc->p_sid;
+         proc->tty_nr = kproc->p_tdev;
+         proc->pgrp = kproc->p__pgid;
+         proc->st_uid = kproc->p_uid;
+         proc->starttime_ctime = kproc->p_ustart_sec;
+         Process_fillStarttimeBuffer(proc);
+         proc->user = UsersTable_getRef(this->super.usersTable, proc->st_uid);
+         ProcessList_add(&this->super, proc);
+         proc->comm = NetBSDProcessList_readProcessName(this->kd, kproc, &proc->basenameOffset);
+      } else {
+         if (settings->updateProcessNames) {
+            free(proc->comm);
+            proc->comm = NetBSDProcessList_readProcessName(this->kd, kproc, &proc->basenameOffset);
+         }
+      }
+
+      proc->m_virt = kproc->p_vm_vsize;
+      proc->m_resident = kproc->p_vm_rssize;
+      proc->percent_mem = (proc->m_resident * pageSizeKB) / (double)(this->super.totalMem) * 100.0;
+      proc->percent_cpu = CLAMP(getpcpu(kproc), 0.0, this->super.cpuCount * 100.0);
+      proc->nlwp = kproc->p_nlwps;
+      proc->nice = kproc->p_nice - 20;
+      proc->time = 100 * (kproc->p_rtime_sec + ((kproc->p_rtime_usec + 500000) / 1000000));
+      proc->priority = kproc->p_priority - PZERO;
+
+      struct kinfo_lwp* klwps = kvm_getlwps(this->kd, kproc->p_pid, kproc->p_paddr, sizeof(struct kinfo_lwp), &nlwps);
+
+      switch (kproc->p_realstat) {
+      case SIDL:     proc->state = 'I'; break;
+      case SACTIVE:
+	// We only consider the first LWP with a one of the below states.
+        for (int j = 0; j < nlwps; j++) {
+          if (klwps) {
```

Comment:
in case `klwps` is `NULL`, `proc->state` is not being set

Reply:
Should be fixed in 896dea7